src/crimson/osd/pg_map.h

```diff
@@ -10,12 +10,75 @@
 
 #include "include/types.h"
 #include "crimson/common/type_helpers.h"
+#include "crimson/common/smp_helpers.h"
 #include "crimson/osd/osd_operation.h"
 #include "osd/osd_types.h"
 
 namespace crimson::osd {
 class PG;
 
+/**
+ * PGShardMapping
+ *
+ * Maps pgs to shards.
+ */
+class PGShardMapping {
+public:
+  /// Returns mapping if present, NULL_CORE otherwise
+  core_id_t get_pg_mapping(spg_t pgid) {
+    auto iter = pg_to_core.find(pgid);
+    ceph_assert(iter == pg_to_core.end() || iter->second != NULL_CORE);
+    return iter == pg_to_core.end() ? NULL_CORE : iter->second;
+  }
+
+  /// Returns mapping for pgid, creates new one if it doesn't already exist
+  core_id_t maybe_create_pg(spg_t pgid) {
+    auto [insert_iter, inserted] = pg_to_core.emplace(pgid, NULL_CORE);
+    if (!inserted) {
+      ceph_assert(insert_iter->second != NULL_CORE);
+      return insert_iter->second;
+    } else {
+      auto iter = core_to_num_pgs.begin();
+      auto min_iter = iter++;
+      for (; iter != core_to_num_pgs.end(); ++iter) {
+	if (min_iter->second > iter->second) min_iter = iter;
+      }
+      insert_iter->second = min_iter->first;
```

Comment:
OK, so we want to evenly distribute PGs across cores. How about the variant of `std::min` that takes comparator?

Reply:
Swapped for std::min_element